samza-sql/src/main/java/org/apache/samza/sql/translator/QueryTranslator.java

```diff
@@ -67,28 +76,56 @@
   private final Map<String, DelegatingSystemDescriptor> systemDescriptors;
   private final Map<String, MessageStream<KV<Object, Object>>> inputMsgStreams;
   private final Map<String, OutputStream> outputMsgStreams;
+  private static final Logger LOG = LoggerFactory.getLogger(QueryTranslator.class);
+
 
   private static class OutputMapFunction implements MapFunction<SamzaSqlRelMessage, KV<Object, Object>> {
     private transient SamzaRelConverter samzaMsgConverter;
+    private transient MetricsRegistry metricsRegistry;
+    private transient SamzaHistogram latency; // milli-seconds
+    private transient Counter outputEvents;
     private final String outputTopic;
     private final int queryId;
+    private String queryLogicalId;
 
     OutputMapFunction(String outputTopic, int queryId) {
       this.outputTopic = outputTopic;
       this.queryId = queryId;
+      queryLogicalId = String.format(TranslatorConstants.LOGSQLID_TEMPLATE, queryId);
     }
 
     @Override
     public void init(Context context) {
       TranslatorContext translatorContext =
           ((SamzaSqlApplicationContext) context.getApplicationTaskContext()).getTranslatorContexts().get(queryId);
       this.samzaMsgConverter = translatorContext.getMsgConverter(outputTopic);
+      ContainerContext containerContext = context.getContainerContext();
+      metricsRegistry = containerContext.getContainerMetricsRegistry();
+      latency = new SamzaHistogram(metricsRegistry, queryLogicalId, TranslatorConstants.LATENCY_NAME);
+      outputEvents = metricsRegistry.newCounter(queryLogicalId, TranslatorConstants.OUTPUT_EVENTS_NAME);
+      outputEvents.clear();
     }
 
     @Override
     public KV<Object, Object> apply(SamzaSqlRelMessage message) {
-      return this.samzaMsgConverter.convertToSamzaMessage(message);
+      if (!message.hasMetadata()) {
+        LOG.warn("Output message with no Metadata: " + message);
+        message.setSamzaSqlRelMsgMetadata(new SamzaSqlRelMsgMetadata(Instant.now().toString()));
```

Comment:
If we agree that when SamzaSqlRelMessage is created, event time or current time is always set, then we won't have to add this code. There is only one place where this determination of event time vs current time needs to be done and that is in Rel converter in convertToRelMessage API implementation. And at all other places where SamzaSqlRelMessage is created (like ProjectTranslator, Join and QueryTranslator outputfunction), we get the time from the input rel message.